### DISCUSSION TITLE
Deprecating _source_ of an imported module.

### DISCUSSION THREAD

**Discussion by gbaz** [ghc-proposals/ghc-proposals#489] [Ideas]
This issue came up in the course of a refactor to Cabal to split out only the syntax modules into Cabal-syntax.

Suppose package A exports module M. We wish to split A, so that package B exports M, and A imports B. Now, everyone that used to depend on package A and still wants M must depend both on A and B.  There are some flag and version tricks to make the transition nice. That said, you may want to "temporarily re-export M from A" just to make everyone's life simple. However, there is currently no sort of deprecation flag that says "if you are depending on M re-exported from A, please also make sure to depend on package B, but if you are depending on the same M, but as originally re-exported from A, you're fine".

I'm not sure at what level of tooling we could detect this, or have a user add this warning, but it seems like it would be perhaps useful as a feature for a variety of cases.

**Comment by Ericson2314** [comment]
It's a good question!

There's a lot of nice things we could be doing to make deprecations / "migrations" easier. "base shims" and backpack are decently unexplored for this sort of thing.,

Ignoring the generalizations, we want a less awkward way to do
```haskell
{-# LANGUAGE PackageImports #-}
module Foo {#-DEPRECATED ... #-} (module Bar) where
import "Cabal-Syntax" Foo as Bar
```
that doesn't make an ambiguous import if both packages are depended upon?

(edit, forgot the deprecation, the whole point of writing it out. Added since.)

**Reply by Ericson2314** [reply]
@gbaz Oops, I forgot to do the deprecation above, which was the whole point of writing it long hand.

Then, except for creating a clashing definition, it should be what you want?

**Reply by gbaz** [reply]
Ok, note that we have a way to do the thing I described avoiding clashes, except without deprecations. It is called reexported modules and I linked to it below. I don't see why we want to instead try to do something else with ghc level imports and reexports.

**Reply by Ericson2314** [reply]
Yeah I was just trying to define what the thing should do. I don't want to do actually have to write that. I do want to use the package conf file.

**Comment by gbaz** [comment]
Nope. I want to be able to have deprecations on reexported modules (https://cabal.readthedocs.io/en/3.4/cabal-package.html#pkg-field-library-reexported-modules).

So in the above, if B exports module M and A re-exports M from B, then if you depend on A alone and use M you get a warning that you should also depend on B, but if you depend on both A and B and use M then you get no such warning.

**Comment by Ericson2314** [comment]
I am reminded of https://gitlab.haskell.org/ghc/ghc/-/issues/4879 / https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0134-deprecating-exports-proposal.rst In particular Simon Marlow's comment from a decade ago: https://gitlab.haskell.org/ghc/ghc/-/issues/4879#note_48067

That seems exactly analogous except this issues is about modules and packages, rather than definitions and modules, right?

**Reply by gbaz** [reply]
Yes, its very analogous. I'm glad to see that proposal was accepted. Less glad that it seems that since then nobody has done any work towards implementing it :-/

**Comment by andreasabel** [comment]
A recent case for this is the `mtl-2.3` misery:
- `mtl-2.3-rc4` removes `ErrorT`, `ListT`: folks have been prepared to this by deprecation warnings for years, so there is little breakage for that (or only breakage where _you have been told so_).
- `mtl-2.3-rc3` also removes reexports with huge breakage in the ecosystem, and there has been no way to warn the customers about importing from the "wrong" module.  In fact, I suspect warning `unused-imports` nudges folks even in the wrong way.

See:
- https://github.com/haskell/mtl/issues/101
- https://github.com/haskell/mtl/issues/102

Cross-posted on: https://gitlab.haskell.org/ghc/ghc/-/issues/4879#note_411752

**Comment by goldfirere** [comment]
I think a deprecated re-export is a fine idea.

**Reply by gbaz** [reply]
Do you have any ideas of how this would work concretely on a module level? (On a function/identifier level https://github.com/ghc-proposals/ghc-proposals/pull/134 seems to suffice were someone to implement it).

I believe that GHC would need some way of being informed that a module is deprecated so it could issue warnings at perhaps the name resolution phase? So that would need to go somewhere -- a package manifest? And in turn cabal would need some syntax of specifying such a thing, such that it could then be packed appropriately for GHC consumption.

**Reply by goldfirere** [reply]
A re-exported module isn't quite a thing. Instead, saying `module A` in an export list exports all identifiers `x` such that both `x` and `A.x` are in scope. (Yes, it really is that strange, as per the Report.) I would just interpret `{-# DEPRECATED "blah" #-} module A` as deprecating each identifier exported. Then, in importing modules, as GHC is building its in-scope table, it labels an identifier as deprecated if all of the imports of that identifier are deprecated. (If the deprecation messages are different, I propose one is just chosen arbitrarily.)

One question, if a deprecated import is re-exported by a `module` export item, does that trigger the warning? Or would it have to be a concrete usage of the identifier. Maybe this is where the water gets murky.

**Reply by Ericson2314** [reply]
@goldfirere @gbaz means reexported at the *package* layer. Not `module Foo (module Bar) where`.

**Reply by Ericson2314** [reply]
Package are, well, meta modules, and the larger lesson here is many thing we need to implement for one we need to implement for the other.

**Reply by gbaz** [reply]
Right. cf https://cabal.readthedocs.io/en/3.4/cabal-package.html#pkg-field-library-reexported-modules for the description of how this works. I don't know how this works in any detail at the GHC layer. Maybe @bgamari or @ezyang does.

**Reply by Ericson2314** [reply]
It is more stuff in the package conf file. See https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/packages.html and `reexported-module`.

**Comment by Ericson2314** [comment]
I wrote up an issue for the GHC aspects of this in https://gitlab.haskell.org/ghc/ghc/-/issues/22489
